semantic_cache/semantic_caching.py

```diff
@@ -0,0 +1,200 @@
+import hashlib
+import json
+from typing import List, Optional, Union
+from upstash_vector import Index
+from langchain_core.outputs.generation import Generation
+
+
+class SemanticCache:
+    """
+    A class to represent a semantic cache using Upstash Vector Database.
+
+    Attributes:
+        min_proximity (float): The minimum proximity score to consider a cache hit.
+        index (Index): The Upstash Vector index object.
+    """
+
+    def __init__(self, url: str, token: str, min_proximity: float = 0.9) -> None:
+        """
+        Initializes the SemanticCache with the given URL and token.
+
+        Args:
+            url (str): The URL of the Upstash Vector database.
+            token (str): The token for accessing the Upstash Vector database.
+            min_proximity (float): The minimum proximity score to consider a cache hit.
+        """
+        self.min_proximity = min_proximity
+        self.index = Index(url=url, token=token)
+
+    def get(self, key: str) -> Optional[str]:
+        """
+        Searches the cache for the key and returns the value if it exists.
+
+        Args:
+            key (str): The key to search in the cache.
+
+        Returns:
+            Optional[str]: The value associated with the key if it exists and meets the proximity score; otherwise, None.
+        """
+        response = self.query_key(key)
+        if response is None or response.score <= self.min_proximity:
+            return None
+        return response.metadata["data"]
+
+    def lookup(
```

Comment:
Is this a public API? Also, why do we dealing with `Generation`s?

Reply:
We wanted the algorthm to be compatible with Langchain, and the Langchain API returns a list of Genaration objects as a llm response

Reply:
Hmm, I see. Did we consider the other alternative: Do not expose these methods here, (and don't depend on langchain), but send a PR to the langchain repo so that we would add our semantic cache to their repo. So, if somebody wants to use Upstash for semantic caching langchain they would import something like `langchain_community.cache.UpstashSemanticCache`.

I am fine with the current API is well, but wondering what you (and the DX team in general) feels about it.

Reply:
I am sorry I missed this one. I think we did not consider this option. Not exposing these methods eliminates the redundant dependency for those who won't use our semantic caching with Langchain. But Langchain compatibility is also important for us, so I am not sure about which option should be taken